### ISSUE TITLE
flang-new crash

### ISSUE THREAD

**Issue by eschnett** [llvm/llvm-project#60342]
I installed `flang-new` from LLVM 15.0.7 via Spack on macOS. It crashes on a particular input, asking me to submit this report.

My shell command is:
```sh
$ ~/src/spack/opt/spack/darwin-ventura-skylake/apple-clang-14.0.0/llvm-15.0.7-wbpxh3adhknrij5evfqp5g6xhus7iqvt/bin/flang-new -DGCC_ATTRIBUTES_NO_ARG_CHECK= -S mpi.F90
```

The error message is:
```console
$ ~/src/spack/opt/spack/darwin-ventura-skylake/apple-clang-14.0.0/llvm-15.0.7-wbpxh3adhknrij5evfqp5g6xhus7iqvt/bin/flang-new -DGCC_ATTRIBUTES_NO_ARG_CHECK= -S mpi.F90
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /Users/eschnett/src/spack/opt/spack/darwin-ventura-skylake/apple-clang-14.0.0/llvm-15.0.7-wbpxh3adhknrij5evfqp5g6xhus7iqvt/bin/flang-new -fc1 -triple x86_64-apple-macosx13.0.0 -S -D GCC_ATTRIBUTES_NO_ARG_CHECK= -fcolor-diagnostics -o mpi.s -x f95-cpp-input mpi.F90
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  flang-new                0x000000010702ac77 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  flang-new                0x0000000107029b4a llvm::sys::RunSignalHandlers() + 250
2  flang-new                0x000000010702b300 SignalHandler(int) + 272
3  libsystem_platform.dylib 0x00007ff81c011c1d _sigtramp + 29
4  libsystem_platform.dylib 0x000000011138b000 _sigtramp + 18446603374694994944
5  flang-new                0x0000000107ad8bf5 void Fortran::parser::Walk<Fortran::parser::CompilerDirective, (anonymous namespace)::PFTBuilder>(Fortran::common::Indirection<Fortran::parser::CompilerDirective, false> const&, (anonymous namespace)::PFTBuilder&) + 197
6  flang-new                0x0000000107ad1e9e std::__1::enable_if<TupleTrait<Fortran::parser::SpecificationPart>, void>::type Fortran::parser::Walk<Fortran::parser::SpecificationPart, (anonymous namespace)::PFTBuilder>(Fortran::parser::SpecificationPart const&, (anonymous namespace)::PFTBuilder&) + 302
7  flang-new                0x0000000107adb1b8 decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<0ul>::__dispatch<std::__1::__variant_detail::__visitation::__variant::__value_visitor<void Fortran::parser::Walk<(anonymous namespace)::PFTBuilder, Fortran::parser::InterfaceBody, Fortran::parser::Statement<Fortran::parser::ProcedureStmt>>(std::__1::variant<Fortran::parser::InterfaceBody, Fortran::parser::Statement<Fortran::parser::ProcedureStmt>> const&, (anonymous namespace)::PFTBuilder&)::'lambda'((anonymous namespace)::PFTBuilder const&)>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Fortran::parser::InterfaceBody, Fortran::parser::Statement<Fortran::parser::ProcedureStmt>> const&>((anonymous namespace)::PFTBuilder, Fortran::parser::InterfaceBody, Fortran::parser::Statement<Fortran::parser::ProcedureStmt>) + 56
8  flang-new                0x0000000107ad9d58 decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<3ul>::__dispatch<std::__1::__variant_detail::__visitation::__variant::__value_visitor<void Fortran::parser::Walk<(anonymous namespace)::PFTBuilder, Fortran::common::Indirection<Fortran::parser::DerivedTypeDef, false>, Fortran::common::Indirection<Fortran::parser::EnumDef, false>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::GenericStmt, false>>, Fortran::common::Indirection<Fortran::parser::InterfaceBlock, false>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::ParameterStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::OldParameterStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::ProcedureDeclarationStmt, false>>, Fortran::parser::Statement<Fortran::parser::OtherSpecificationStmt>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::TypeDeclarationStmt, false>>, Fortran::common::Indirection<Fortran::parser::StructureDef, false>, Fortran::common::Indirection<Fortran::parser::OpenACCDeclarativeConstruct, false>, Fortran::common::Indirection<Fortran::parser::OpenMPDeclarativeConstruct, false>, Fortran::common::Indirection<Fortran::parser::CompilerDirective, false>>(std::__1::variant<Fortran::common::Indirection<Fortran::parser::DerivedTypeDef, false>, Fortran::common::Indirection<Fortran::parser::EnumDef, false>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::GenericStmt, false>>, Fortran::common::Indirection<Fortran::parser::InterfaceBlock, false>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::ParameterStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::OldParameterStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::ProcedureDeclarationStmt, false>>, Fortran::parser::Statement<Fortran::parser::OtherSpecificationStmt>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::TypeDeclarationStmt, false>>, Fortran::common::Indirection<Fortran::parser::StructureDef, false>, Fortran::common::Indirection<Fortran::parser::OpenACCDeclarativeConstruct, false>, Fortran::common::Indirection<Fortran::parser::OpenMPDeclarativeConstruct, false>, Fortran::common::Indirection<Fortran::parser::CompilerDirective, false>> const&, (anonymous namespace)::PFTBuilder&)::'lambda'((anonymous namespace)::PFTBuilder const&)>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Fortran::common::Indirection<Fortran::parser::DerivedTypeDef, false>, Fortran::common::Indirection<Fortran::parser::EnumDef, false>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::GenericStmt, false>>, Fortran::common::Indirection<Fortran::parser::InterfaceBlock, false>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::ParameterStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::OldParameterStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::ProcedureDeclarationStmt, false>>, Fortran::parser::Statement<Fortran::parser::OtherSpecificationStmt>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::TypeDeclarationStmt, false>>, Fortran::common::Indirection<Fortran::parser::StructureDef, false>, Fortran::common::Indirection<Fortran::parser::OpenACCDeclarativeConstruct, false>, Fortran::common::Indirection<Fortran::parser::OpenMPDeclarativeConstruct, false>, Fortran::common::Indirection<Fortran::parser::CompilerDirective, false>> const&>((anonymous namespace)::PFTBuilder, Fortran::common::Indirection<Fortran::parser::DerivedTypeDef, false>, Fortran::common::Indirection<Fortran::parser::EnumDef, false>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::GenericStmt, false>>, Fortran::common::Indirection<Fortran::parser::InterfaceBlock, false>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::ParameterStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::OldParameterStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::ProcedureDeclarationStmt, false>>, Fortran::parser::Statement<Fortran::parser::OtherSpecificationStmt>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::TypeDeclarationStmt, false>>, Fortran::common::Indirection<Fortran::parser::StructureDef, false>, Fortran::common::Indirection<Fortran::parser::OpenACCDeclarativeConstruct, false>, Fortran::common::Indirection<Fortran::parser::OpenMPDeclarativeConstruct, false>, Fortran::common::Indirection<Fortran::parser::CompilerDirective, false>) + 152
9  flang-new                0x0000000107ad9865 decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<0ul>::__dispatch<std::__1::__variant_detail::__visitation::__variant::__value_visitor<void Fortran::parser::Walk<(anonymous namespace)::PFTBuilder, Fortran::parser::SpecificationConstruct, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::DataStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::FormatStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::EntryStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::StmtFunctionStmt, false>>, Fortran::parser::ErrorRecovery>(std::__1::variant<Fortran::parser::SpecificationConstruct, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::DataStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::FormatStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::EntryStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::StmtFunctionStmt, false>>, Fortran::parser::ErrorRecovery> const&, (anonymous namespace)::PFTBuilder&)::'lambda'((anonymous namespace)::PFTBuilder const&)>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Fortran::parser::SpecificationConstruct, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::DataStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::FormatStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::EntryStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::StmtFunctionStmt, false>>, Fortran::parser::ErrorRecovery> const&>((anonymous namespace)::PFTBuilder, Fortran::parser::SpecificationConstruct, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::DataStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::FormatStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::EntryStmt, false>>, Fortran::parser::Statement<Fortran::common::Indirection<Fortran::parser::StmtFunctionStmt, false>>, Fortran::parser::ErrorRecovery) + 53
10 flang-new                0x0000000107ad1ef4 std::__1::enable_if<TupleTrait<Fortran::parser::SpecificationPart>, void>::type Fortran::parser::Walk<Fortran::parser::SpecificationPart, (anonymous namespace)::PFTBuilder>(Fortran::parser::SpecificationPart const&, (anonymous namespace)::PFTBuilder&) + 388
11 flang-new                0x0000000107ace2a9 decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<3ul>::__dispatch<std::__1::__variant_detail::__visitation::__variant::__value_visitor<void Fortran::parser::Walk<(anonymous namespace)::PFTBuilder, Fortran::common::Indirection<Fortran::parser::MainProgram, false>, Fortran::common::Indirection<Fortran::parser::FunctionSubprogram, false>, Fortran::common::Indirection<Fortran::parser::SubroutineSubprogram, false>, Fortran::common::Indirection<Fortran::parser::Module, false>, Fortran::common::Indirection<Fortran::parser::Submodule, false>, Fortran::common::Indirection<Fortran::parser::BlockData, false>, Fortran::common::Indirection<Fortran::parser::CompilerDirective, false>>(std::__1::variant<Fortran::common::Indirection<Fortran::parser::MainProgram, false>, Fortran::common::Indirection<Fortran::parser::FunctionSubprogram, false>, Fortran::common::Indirection<Fortran::parser::SubroutineSubprogram, false>, Fortran::common::Indirection<Fortran::parser::Module, false>, Fortran::common::Indirection<Fortran::parser::Submodule, false>, Fortran::common::Indirection<Fortran::parser::BlockData, false>, Fortran::common::Indirection<Fortran::parser::CompilerDirective, false>> const&, (anonymous namespace)::PFTBuilder&)::'lambda'((anonymous namespace)::PFTBuilder const&)>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, Fortran::common::Indirection<Fortran::parser::MainProgram, false>, Fortran::common::Indirection<Fortran::parser::FunctionSubprogram, false>, Fortran::common::Indirection<Fortran::parser::SubroutineSubprogram, false>, Fortran::common::Indirection<Fortran::parser::Module, false>, Fortran::common::Indirection<Fortran::parser::Submodule, false>, Fortran::common::Indirection<Fortran::parser::BlockData, false>, Fortran::common::Indirection<Fortran::parser::CompilerDirective, false>> const&>((anonymous namespace)::PFTBuilder, Fortran::common::Indirection<Fortran::parser::MainProgram, false>, Fortran::common::Indirection<Fortran::parser::FunctionSubprogram, false>, Fortran::common::Indirection<Fortran::parser::SubroutineSubprogram, false>, Fortran::common::Indirection<Fortran::parser::Module, false>, Fortran::common::Indirection<Fortran::parser::Submodule, false>, Fortran::common::Indirection<Fortran::parser::BlockData, false>, Fortran::common::Indirection<Fortran::parser::CompilerDirective, false>) + 217
12 flang-new                0x0000000107a9d33b Fortran::lower::createPFT(Fortran::parser::Program const&, Fortran::semantics::SemanticsContext const&) + 299
13 flang-new                0x0000000107654901 Fortran::lower::LoweringBridge::lower(Fortran::parser::Program const&, Fortran::semantics::SemanticsContext const&) + 49
14 flang-new                0x0000000107037990 Fortran::frontend::CodeGenAction::beginSourceFileAction() + 3360
15 flang-new                0x0000000107035a79 Fortran::frontend::FrontendAction::beginSourceFile(Fortran::frontend::CompilerInstance&, Fortran::frontend::FrontendInputFile const&) + 297
16 flang-new                0x000000010702d44f Fortran::frontend::CompilerInstance::executeAction(Fortran::frontend::FrontendAction&) + 127
17 flang-new                0x000000010710af71 Fortran::frontend::executeCompilerInvocation(Fortran::frontend::CompilerInstance*) + 2449
18 flang-new                0x0000000106bb49f9 fc1_main(llvm::ArrayRef<char const*>, char const*) + 505
19 flang-new                0x0000000106bb30db main + 651
20 dyld                     0x00007ff81bcb4310 start + 2432
flang-new: error: unable to execute command: Segmentation fault: 11
flang-new: error: flang frontend command failed due to signal (use -v to see invocation)
flang-new version 15.0.7
Target: x86_64-apple-darwin22.2.0
Thread model: posix
InstalledDir: /Users/eschnett/src/spack/opt/spack/darwin-ventura-skylake/apple-clang-14.0.0/llvm-15.0.7-wbpxh3adhknrij5evfqp5g6xhus7iqvt/bin
flang-new: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
flang-new: note: diagnostic msg: /var/folders/gb/x5lhfpj15ln66g6br549tz5m0000gs/T/mpi-82108d
flang-new: note: diagnostic msg: /var/folders/gb/x5lhfpj15ln66g6br549tz5m0000gs/T/mpi-82108d.sh
flang-new: note: diagnostic msg: Crash backtrace is located in
flang-new: note: diagnostic msg: /Users/eschnett/Library/Logs/DiagnosticReports/flang-new_<YYYY-MM-DD-HHMMSS>_<hostname>.crash
flang-new: note: diagnostic msg: (choose the .crash file that corresponds to your crash)
flang-new: note: diagnostic msg:

********************
```

[mpi-82108d.sh.txt](https://github.com/llvm/llvm-project/files/1052262
[mpi-82108d.txt](https://github.com/llvm/llvm-project/files/10522629/mp
[flang-new-2023-01-27-144820.ips.txt](https://github.com/llvm/llvm-project/files/10522632/flang-new-2023-01-27-144820.ips.txt)
i-82108d.txt)
5/mpi-82108d.sh.txt)

**Comment by clementval** [comment]
Looks like smth related to compiler directive. Can you attach your input file?

**Comment by eschnett** [comment]
The input file is `mpi.F90`. It includes a few other files. I am attaching them all. (They are renamed to have a `.txt` suffix because Github makes me do so.)

[mpi.F90.txt](https://github.com/llvm/llvm-project/files/10524332/mpi.F90.txt)
[mpi_decl_constants_fortran.h.txt](https://github.com/llvm/llvm-project/files/10524333/mpi_decl_constants_fortran.h.txt)
[mpiabif.h.txt](https://github.com/llvm/llvm-project/files/10524334/mpiabif.h
[mpif_version.h.txt](https://github.com/llvm/llvm-project/files/10524335/mpif_version.h.txt)
.txt)

**Comment by eschnett** [comment]
Yes, there are directives in the file `mpi.F90`:

```Fortran
       !dir$ ignore_tkr(tkr) buf
       !gcc$ attributes no_arg_check :: buf
       integer, intent(in) :: buf(*)
```

(The `GCC_ATTRIBUTES_NO_ARG_CHECK` in the source code is removed via the preprocessor.)

**Comment by clementval** [comment]
The crash is coming from the directive. This small code snippet can reproduce it. 

```
module mpi
  implicit none
  public

  interface
     subroutine mpi_send(buf, count, datatype, dest, tag, comm, ierror)
       implicit none
       !dir$ ignore_tkr(tkr) buf
       integer, intent(in) :: buf(*)
       integer, intent(in) :: count
       integer, intent(in) :: datatype
       integer, intent(in) :: dest
       integer, intent(in) :: tag
       integer, intent(in) :: comm
       integer, intent(out) :: ierror
     end subroutine mpi_send
  end interface

end module mpi
```

**Comment by eschnett** [comment]
I assumed that the `dir` directive would be ignored. There is an equivalent `gcc` directive nearby for gfortran (not present in this reduced example), and I'd be happy to add a flang-specific directive as well if necessary.

**Comment by clementval** [comment]
It should be ignored as it is not currently supported but it seems to trigger the error downstream.

**Comment by vdonaldson** [comment]
Lowering code currently allows for directives inside a program or subprogram, or outside any program unit. This case of a directive in the specification part of a module is neither of those, but should also be allowed. 
```
module mm
  interface
     subroutine ss(buf)
       !dir$ ignore_tkr(tkr) buf
       integer :: buf(*)
     end subroutine ss
  end interface
end module
```
Except for OpenMP directives, most directives are currently ignored, so this code should generate an `ignoring all compiler directives` message.
